File: vbnet/acroform/form_fields/form_fields.vb
' form_fields -- VB.NET port of examples\Vb6\acroform\form_fields
Imports System
Imports System.IO
Imports LumasPdfSdk

Module FormFields
    Private errDel As TErrorProc

    Public Function PDFError(ByVal Data As IntPtr, ByVal ErrCode As Integer, ByVal ErrMessage As String, ByVal ErrType As Integer) As Integer
        Return 0
    End Function

    Sub Main()
        Dim f As Integer, r As Integer
        Dim y As Double
        Dim pdf As IntPtr = LumasPdf.pdfNewPDF()
        errDel = New TErrorProc(AddressOf PDFError)
        LumasPdf.pdfSetOnErrorProc(pdf, IntPtr.Zero, errDel)
        LumasPdf.pdfCreateNewPDFW(pdf, "")

        LumasPdf.pdfSetPageCoords(pdf, TPageCoord.pcTopDown)

        LumasPdf.pdfAppend(pdf)
        y = 50.0
        LumasPdf.pdfSetFontW(pdf, "Helvetica", LumasPdfConsts.fsRegular, 10.0, False, TCodepage.cp1252)
        LumasPdf.pdfWriteTextW(pdf, 50.0, y, "Text fields:")

        y = y + 15.0
        f = LumasPdf.pdfCreateTextField(pdf, "Text1", -1, 0, 0, 50.0, y, 200.0, 20.0)
        LumasPdf.pdfSetTextFieldValueW(pdf, f, "", "Single line text...", LumasPdfConsts.taLeft)

        y = y + 30.0
        f = LumasPdf.pdfCreateTextField(pdf, "Text2", -1, 1, 0, 50.0, y, 200.0, 50.0)
        LumasPdf.pdfSetTextFieldValueW(pdf, f, "", "This field accepts multi-line text. The maximum text length can be restricted if necessary.", LumasPdfConsts.taLeft)

        y = y + 60.0
        LumasPdf.pdfWriteTextW(pdf, 50.0, y, "A password field:")
        y = y + 15.0
        f = LumasPdf.pdfCreateTextField(pdf, "Text3", -1, 0, 0, 50.0, y, 200.0, 20.0)
        LumasPdf.pdfSetFieldFlags(pdf, f, LumasPdfConsts.ffPassword, False)
        LumasPdf.pdfSetTextFieldValueW(pdf, f, "", "**********", LumasPdfConsts.taLeft)

        y = y + 30.0
        LumasPdf.pdfWriteTextW(pdf, 50.0, y, "A fixed length field separated into combs:")
        y = y + 15.0
        f = LumasPdf.pdfCreateTextField(pdf, "Text4", -1, 0, 10, 50.0, y, 200.0, 20.0)
        LumasPdf.pdfSetFieldFlags(pdf, f, LumasPdfConsts.ffComb, False)

        y = 50.0
        LumasPdf.pdfWriteTextW(pdf, 350.0, y, "Choice fields:")
        y = y + 15.0
        f = LumasPdf.pdfCreateComboBox(pdf, "Combo1", 1, -1, 350.0, y, 200.0, 20.0)
        LumasPdf.pdfAddValToChoiceFieldW(pdf, f, "", " Select a value...", 1)
        LumasPdf.pdfAddValToChoiceFieldW(pdf, f, "Apple", "Apple", 0)
        LumasPdf.pdfAddValToChoiceFieldW(pdf, f, "Banana", "Banana", 0)
        LumasPdf.pdfAddValToChoiceFieldW(pdf, f, "Pear", "Pear", 0)
        LumasPdf.pdfAddValToChoiceFieldW(pdf, f, "Grape", "Grape", 0)
        LumasPdf.pdfAddValToChoiceFieldW(pdf, f, "Orange", "Orange", 0)

        y = y + 30.0
        f = LumasPdf.pdfCreateListBox(pdf, "List", True, -1, 350.0, y, 200.0, 50.0)
        LumasPdf.pdfAddValToChoiceFieldW(pdf, f, "Apple", "Apple", 0)
        LumasPdf.pdfAddValToChoiceFieldW(pdf, f, "Banana", "Banana", 1)
        LumasPdf.pdfAddValToChoiceFieldW(pdf, f, "Pear", "Pear", 0)
        LumasPdf.pdfAddValToChoiceFieldW(pdf, f, "Grape", "Grape", 0)
        LumasPdf.pdfAddValToChoiceFieldW(pdf, f, "Orange", "Orange", 0)

        y = y + 60.0
        LumasPdf.pdfWriteTextW(pdf, 350.0, y, "Editable combo box:")
        y = y + 15.0
        f = LumasPdf.pdfCreateComboBox(pdf, "Combo2", 1, -1, 350.0, y, 200.0, 20.0)
        LumasPdf.pdfAddValToChoiceFieldW(pdf, f, "Apple", "Apple", 0)
        LumasPdf.pdfAddValToChoiceFieldW(pdf, f, "Banana", "Banana", 0)
        LumasPdf.pdfAddValToChoiceFieldW(pdf, f, "Pear", "Pear", 0)
        LumasPdf.pdfAddValToChoiceFieldW(pdf, f, "Grape", "Grape", 0)
        LumasPdf.pdfAddValToChoiceFieldW(pdf, f, "Orange", "Orange", 0)
        LumasPdf.pdfSetFieldFlags(pdf, f, LumasPdfConsts.ffEdit, False)
        LumasPdf.pdfSetFieldExpValueW(pdf, f, 1000, "Select or enter a value...", "", True)

        y = y + 30.0
        LumasPdf.pdfWriteTextW(pdf, 350.0, y, "Check boxes / Radio buttons:")

        y = y + 15.0
        LumasPdf.pdfChangeFontSize(pdf, 1.0)
        LumasPdf.pdfCreateCheckBox(pdf, "N1", "C1", 1, -1, 350.0, y, 20.0, 20.0)
        LumasPdf.pdfCreateCheckBox(pdf, "N2", "C2", 1, -1, 380.0, y, 20.0, 20.0)
        LumasPdf.pdfCreateCheckBox(pdf, "N3", "C1", 1, -1, 410.0, y, 20.0, 20.0)

        LumasPdf.pdfCreateCheckBox(pdf, "G1", "C1", 0, -1, 450.0, y, 20.0, 20.0)
        LumasPdf.pdfCreateCheckBox(pdf, "G1", "C2", 0, -1, 480.0, y, 20.0, 20.0)
        LumasPdf.pdfCreateCheckBox(pdf, "G1", "C1", 1, -1, 510.0, y, 20.0, 20.0)
        LumasPdf.pdfCreateCheckBox(pdf, "G1", "C2", 0, -1, 540.0, y, 20.0, 20.0)

        y = y + 30.0
        LumasPdf.pdfChangeFontSize(pdf, 15.0)
        LumasPdf.pdfSetCheckBoxChar(pdf, TCheckBoxChar.ccCircle)
        r = LumasPdf.pdfCreateRadioButton(pdf, "Radio1", "R1", 1, -1, 350.0, y, 20.0, 20.0)
        LumasPdf.pdfSetCheckBoxDefState(pdf, r, False)
        LumasPdf.pdfCreateCheckBox(pdf, "", "R2", 0, r, 380.0, y, 20.0, 20.0)
        LumasPdf.pdfCreateCheckBox(pdf, "", "R3", 0, r, 410.0, y, 20.0, 20.0)

        r = LumasPdf.pdfCreateRadioButton(pdf, "Radio2", "R1", 1, -1, 450.0, y, 20.0, 20.0)
        LumasPdf.pdfSetFieldFlags(pdf, r, LumasPdfConsts.ffRadioIsUnion, False)
        LumasPdf.pdfCreateCheckBox(pdf, "", "R2", 0, r, 480.0, y, 20.0, 20.0)
        LumasPdf.pdfCreateCheckBox(pdf, "", "R1", 1, r, 510.0, y, 20.0, 20.0)
        LumasPdf.pdfCreateCheckBox(pdf, "", "R2", 0, r, 540.0, y, 20.0, 20.0)
        LumasPdf.pdfEndPage(pdf)

        If LumasPdf.pdfHaveOpenDoc(pdf) <> 0 Then
            Dim outFile As String = Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "out.pdf")
            If LumasPdf.pdfOpenOutputFileW(pdf, outFile) = 0 Then
                LumasPdf.pdfDeletePDF(pdf)
                Return
            End If
            If LumasPdf.pdfCloseFile(pdf) <> 0 Then
                Console.WriteLine("PDF file """ & outFile & """ successfully created!")
            End If
        End If

        LumasPdf.pdfDeletePDF(pdf)
    End Sub
End Module

File: python/acroform/form_fields/form_fields.py
# ============================================================================
#  form_fields -- Python (ctypes) port of examples\Vb6\acroform\form_fields\form_fields.bas
#  Text fields (single/multi-line/password/comb), combo boxes, list boxes,
#  editable combo, check boxes and radio buttons.
# ============================================================================
import os
import sys
# the wrapper lives at <root>/wrappers/python (source checkout) or beside
# the example tree (shipped package) -- find it without a hard-coded path
_d = os.path.dirname(os.path.abspath(__file__))
for _ in range(6):
    for _c in (os.path.join(_d, 'wrappers', 'python'), _d):
        if os.path.isfile(os.path.join(_c, 'lumaspdf.py')):
            sys.path.insert(0, _c)
            break
    else:
        _d = os.path.dirname(_d)
        continue
    break
import lumaspdf as L


@L.TErrorProc
def err_proc(data, errcode, errmsg, errtype):
    return 0


def main():
    pdf = L.pdfNewPDF()
    L.pdfSetOnErrorProc(pdf, 0, err_proc)
    L.pdfCreateNewPDFA(pdf, b"")

    L.pdfSetPageCoords(pdf, L.pcTopDown)

    L.pdfAppend(pdf)
    y = 50.0
    L.pdfSetFontA(pdf, b"Helvetica", L.fsRegular, 10.0, 0, L.cp1252)
    L.pdfWriteTextA(pdf, 50.0, y, b"Text fields:")

    y = y + 15.0
    f = L.pdfCreateTextField(pdf, b"Text1", -1, 0, 0, 50.0, y, 200.0, 20.0)
    L.pdfSetTextFieldValueA(pdf, f, b"", b"Single line text...", L.taLeft)

    y = y + 30.0
    f = L.pdfCreateTextField(pdf, b"Text2", -1, 1, 0, 50.0, y, 200.0, 50.0)
    L.pdfSetTextFieldValueA(pdf, f, b"", b"This field accepts multi-line text. The maximum text length can be restricted if necessary.", L.taLeft)

    y = y + 60.0
    L.pdfWriteTextA(pdf, 50.0, y, b"A password field:")
    y = y + 15.0
    f = L.pdfCreateTextField(pdf, b"Text3", -1, 0, 0, 50.0, y, 200.0, 20.0)
    L.pdfSetFieldFlags(pdf, f, L.ffPassword, 0)
    L.pdfSetTextFieldValueA(pdf, f, b"", b"**********", L.taLeft)

    y = y + 30.0
    L.pdfWriteTextA(pdf, 50.0, y, b"A fixed length field separated into combs:")
    y = y + 15.0
    f = L.pdfCreateTextField(pdf, b"Text4", -1, 0, 10, 50.0, y, 200.0, 20.0)
    L.pdfSetFieldFlags(pdf, f, L.ffComb, 0)

    y = 50.0
    L.pdfWriteTextA(pdf, 350.0, y, b"Choice fields:")
    y = y + 15.0
    f = L.pdfCreateComboBox(pdf, b"Combo1", 1, -1, 350.0, y, 200.0, 20.0)
    L.pdfAddValToChoiceFieldA(pdf, f, b"", b" Select a value...", 1)
    L.pdfAddValToChoiceFieldA(pdf, f, b"Apple", b"Apple", 0)
    L.pdfAddValToChoiceFieldA(pdf, f, b"Banana", b"Banana", 0)
    L.pdfAddValToChoiceFieldA(pdf, f, b"Pear", b"Pear", 0)
    L.pdfAddValToChoiceFieldA(pdf, f, b"Grape", b"Grape", 0)
    L.pdfAddValToChoiceFieldA(pdf, f, b"Orange", b"Orange", 0)

    y = y + 30.0
    f = L.pdfCreateListBox(pdf, b"List", 1, -1, 350.0, y, 200.0, 50.0)
    L.pdfAddValToChoiceFieldA(pdf, f, b"Apple", b"Apple", 0)
    L.pdfAddValToChoiceFieldA(pdf, f, b"Banana", b"Banana", 1)
    L.pdfAddValToChoiceFieldA(pdf, f, b"Pear", b"Pear", 0)
    L.pdfAddValToChoiceFieldA(pdf, f, b"Grape", b"Grape", 0)
    L.pdfAddValToChoiceFieldA(pdf, f, b"Orange", b"Orange", 0)

    y = y + 60.0
    L.pdfWriteTextA(pdf, 350.0, y, b"Editable combo box:")
    y = y + 15.0
    f = L.pdfCreateComboBox(pdf, b"Combo2", 1, -1, 350.0, y, 200.0, 20.0)
    L.pdfAddValToChoiceFieldA(pdf, f, b"Apple", b"Apple", 0)
    L.pdfAddValToChoiceFieldA(pdf, f, b"Banana", b"Banana", 0)
    L.pdfAddValToChoiceFieldA(pdf, f, b"Pear", b"Pear", 0)
    L.pdfAddValToChoiceFieldA(pdf, f, b"Grape", b"Grape", 0)
    L.pdfAddValToChoiceFieldA(pdf, f, b"Orange", b"Orange", 0)
    L.pdfSetFieldFlags(pdf, f, L.ffEdit, 0)
    L.pdfSetFieldExpValueA(pdf, f, 1000, b"Select or enter a value...", b"", 1)

    y = y + 30.0
    L.pdfWriteTextA(pdf, 350.0, y, b"Check boxes / Radio buttons:")

    y = y + 15.0
    L.pdfChangeFontSize(pdf, 1.0)
    L.pdfCreateCheckBox(pdf, b"N1", b"C1", 1, -1, 350.0, y, 20.0, 20.0)
    L.pdfCreateCheckBox(pdf, b"N2", b"C2", 1, -1, 380.0, y, 20.0, 20.0)
    L.pdfCreateCheckBox(pdf, b"N3", b"C1", 1, -1, 410.0, y, 20.0, 20.0)

    L.pdfCreateCheckBox(pdf, b"G1", b"C1", 0, -1, 450.0, y, 20.0, 20.0)
    L.pdfCreateCheckBox(pdf, b"G1", b"C2", 0, -1, 480.0, y, 20.0, 20.0)
    L.pdfCreateCheckBox(pdf, b"G1", b"C1", 1, -1, 510.0, y, 20.0, 20.0)
    L.pdfCreateCheckBox(pdf, b"G1", b"C2", 0, -1, 540.0, y, 20.0, 20.0)

    y = y + 30.0
    L.pdfChangeFontSize(pdf, 15.0)
    L.pdfSetCheckBoxChar(pdf, L.ccCircle)
    r = L.pdfCreateRadioButton(pdf, b"Radio1", b"R1", 1, -1, 350.0, y, 20.0, 20.0)
    L.pdfSetCheckBoxDefState(pdf, r, 0)
    L.pdfCreateCheckBox(pdf, b"", b"R2", 0, r, 380.0, y, 20.0, 20.0)
    L.pdfCreateCheckBox(pdf, b"", b"R3", 0, r, 410.0, y, 20.0, 20.0)

    r = L.pdfCreateRadioButton(pdf, b"Radio2", b"R1", 1, -1, 450.0, y, 20.0, 20.0)
    L.pdfSetFieldFlags(pdf, r, L.ffRadioIsUnion, 0)
    L.pdfCreateCheckBox(pdf, b"", b"R2", 0, r, 480.0, y, 20.0, 20.0)
    L.pdfCreateCheckBox(pdf, b"", b"R1", 1, r, 510.0, y, 20.0, 20.0)
    L.pdfCreateCheckBox(pdf, b"", b"R2", 0, r, 540.0, y, 20.0, 20.0)
    L.pdfEndPage(pdf)

    if L.pdfHaveOpenDoc(pdf) != 0:
        out_file = os.path.join(os.path.dirname(os.path.abspath(__file__)), "out.pdf")
        if L.pdfOpenOutputFileA(pdf, out_file.encode("latin-1")) == 0:
            L.pdfDeletePDF(pdf)
            return
        if L.pdfCloseFile(pdf) != 0:
            print('PDF file "' + out_file + '" successfully created!')

    L.pdfDeletePDF(pdf)


if __name__ == "__main__":
    main()
